File: hasPathSum.py
class TreeNode:
	def __init__(self, x, l=None, r=None):
		self.val=x
		self.left=l
		self.right=r

class Solution:
	def hasPathSum(self,root,sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
		if not root:
			return False
		if root.val==sum and not root.left and not root.right:
			return True
		return self.hasPathSum(root.left, sum-root.val) or self.hasPathSum(root.right, sum-root.val)

n2=TreeNode(2)
n1=TreeNode(1,n2,None)
s=Solution()
res=s.hasPathSum(n1,1)
print(res)

File: hasPathSum.cc
#include <bits/stdc++.h>
using namespace std;
struct TreeNode {
	int val;
	TreeNode *left, *right;
	TreeNode(int v): val(v), left(nullptr), right(nullptr) {}
	TreeNode(int v, TreeNode* l, TreeNode* r): val(v), left(l), right(r) {}
};
class Solution {
public:
	// dfs
	bool hasPathSum(TreeNode *root, int sum) {
		if (!root) return false;
		int v = root->val;
		if (v==sum && !root->left && !root->right) return true;
		return hasPathSum(root->left, sum-v) || hasPathSum(root->right, sum-v);
	}
};
int main(){
	TreeNode n9(1),n8(2),n7(7),n6(4,nullptr,&n9),n5(13),n4(11,&n7,&n8),n3(8,&n5,&n6),n2(4,&n4,nullptr),n1(5,&n2,&n3);
	TreeNode m2(2),m1(1,&m2,nullptr);
	Solution s;
	cout << s.hasPathSum(&n1, 22) << endl;
	cout << s.hasPathSum(&m1, 1) << endl;
}
